### PR TITLE
HTTP and SSL improvements

### PR DESCRIPTION
**Pull request by beaucollins**
Updates the WebSocket and REST transport layer with AndroidAsync library.

Implements SSL pinning against simperium.com's SPKI.

Fixes some async bugs around database indexing and reporting of download progress during an initial bucket index download.

Uses Android Gradle's build flavors feature to produce a library compatible with testing that includes all of the mock components. Production flavor is named `client` while testing compatible flavor is named `support`.

When using the library from source code reference the flavor and build type to get testing and debugging support:

```
dependencies {
    compile project(path: ":Simperium", configuration:"supportDebug")
}
```

### CODE REVIEW

**Comment by beaucollins** [comment]
Syncing works but has problems reconnecting at the moment.

`MockWebSocketClient` will need to be tossed out and a new interface provided for tests.

**Comment by beaucollins** [comment]
Finally figured out this travis stuff :P

**Comment by beaucollins** [comment]
When Simplenote is built using the pinning code but with a different SPKI it fails to connect (as expected):

```
Error: com.koushikdutta.async.http.ConnectionClosedException: connection closed before response completed.
... stacktrace ...
Caused by: com.koushikdutta.async.AsyncSSLException: Peer not trusted by any of the system trust managers.
```

File: Simperium/src/main/java/com/simperium/android/AsyncAuthClient.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
+package com.simperium.android;
+
+import com.simperium.BuildConfig;
```
Comment (roundhill) [inline Simperium/src/main/java/com/simperium/android/AsyncAuthClient.java]: There's some unused imports here.

File: Simperium/src/main/java/com/simperium/android/AsyncAuthClient.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import org.json.JSONObject;
+
+public class AsyncAuthClient implements AuthProvider {
+
+    public static final String TAG = "Simperium.AsyncAuthClient";
+
+    private static final String AUTH_URL = "https://auth.simperium.com/1";
+    private static final String CREATE_PATH = "create/";
+    private static final String AUTHORIZE_PATH = "authorize/";
+    private static final String JSON_CONTENT_TYPE = "application/json";
```
Comment (roundhill) [inline Simperium/src/main/java/com/simperium/android/AsyncAuthClient.java]: `JSON_CONTENT_TYPE` is unused.

File: Simperium/src/main/java/com/simperium/android/AsyncWebSocketProvider.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
+package com.simperium.android;
+
+import com.simperium.BuildConfig;
+
+import com.koushikdutta.async.callback.CompletedCallback;
```
Comment (roundhill) [inline Simperium/src/main/java/com/simperium/android/AsyncWebSocketProvider.java:3]: Also a few unused imports here.

File: Simperium/src/main/java/com/simperium/android/WebSocketManager.java [hunk 4]
```diff
@@ -5,18 +5,17 @@
...
+
 import com.simperium.client.Bucket;
 import com.simperium.client.Channel;
 import com.simperium.client.ChannelProvider;
 import com.simperium.util.Logger;
 
-import org.apache.http.message.BasicNameValuePair;
-
-import java.net.URI;
 import java.util.Arrays;
```
Comment (roundhill) [inline Simperium/src/main/java/com/simperium/android/WebSocketManager.java]: Unused imports ;)

File: Simperium/src/main/java/com/simperium/android/WebSocketManager.java [hunk 5]
```diff
@@ -30,69 +29,64 @@
...
-    private WebSocketClient socketClient;
-    private boolean reconnect = true;
-    private HashMap<Channel,Integer> channelIndex = new HashMap<Channel,Integer>();
-    private HashMap<Integer,Channel> channels = new HashMap<Integer,Channel>();
-    private URI socketURI;
+    final protected ConnectionProvider mConnectionProvider;
+    protected Connection mConnection = new NullConnection();
+    protected String mSocketURI;
+    private String mAppId, mSessionId;
+    private String mClientId;
```
Comment (roundhill) [inline Simperium/src/main/java/com/simperium/android/WebSocketManager.java]: `mClientId` is never used.

**Comment by roundhill** [comment]
Looks good besides a few things:
- Remove those icon placeholder drawables
- Add getter/setter for `BucketObject.mProperties`

**Comment by roundhill** [comment]
:shipit:

**Pull request merged**